sdb/commands/walk.py

```diff
@@ -44,17 +52,12 @@ def call(self, objs: Iterable[drgn.Object]) -> Iterable[drgn.Object]:
             except StopIteration:
                 continue
 
-            print("The following types have walkers:")
-            print("\t%-20s %-20s" % ("WALKER", "TYPE"))
-            for type_, class_ in baked:
-                print("\t%-20s %-20s" % (class_(self.prog).names[0], type_))
-            raise TypeError("no walker found for input of type {}".format(
-                i.type_))
+            raise sdb.CommandError(
+                self.name,
+                Walk._help_message() + "no walker found for input of type {}".format(
```

Comment:
Would it be better if `_help_message` optionally accepted a "type" parameter, and handled it properly, rather than having the consumer do this string concatenation?

Reply:
Could we change this:
```
sdb> spa | walk
sdb: walk: The following types have walkers:
	WALKER               TYPE
	avl                  avl_tree_t *
	spl_list             list_t *
	zfs_btree            zfs_btree_t *
no walker found for input of type spa_t *
```
to this:
```
sdb> spa | walk
sdb: walk: no walker found for input of type spa_t *
The following types have walkers:
	WALKER               TYPE
	avl                  avl_tree_t *
	spl_list             list_t *
	zfs_btree            zfs_btree_t *
```
IMO, having the error message first helps make the error more understandable.

Reply:
> Would it be better if _help_message optionally accepted a "type" parameter, and handled it properly, rather than having the consumer do this string concatenation?

It seems more or less the same, but doing it the way I did has less exceptions / if-cases.  I can change it if you want.

> print message first

I found it easier to read with the most important info at the end, but I can change it.

Reply:
> It seems more or less the same, but doing it the way I did has less exceptions / if-cases. I can change it if you want.

I think it'd be better to add the type parameter so that the formatting of the message is isolated in the `_help_message` function, and assumptions about the format don't leak into the consumers.

Ultimately, it's not a big deal, so I'll leave it up to you; either way is fine.

> I found it easier to read with the most important info at the end, but I can change it.

I think the error should be first for the following reasons:

1. to be more consistent with other error messages; e.g.
```
sdb> metaslab
sdb: metaslab: command requires an input
```
```
sdb> spa | walk
sdb: walk: no walker found for input of type spa_t *
```
vs.
```
sdb> spa | walk
sdb: walk: The following types have walkers:
```

2. when reading the entire message top to bottom, IMO, it "flows" better; i.e. 1. here's the error, 2. here's is potentially how to fix it.. vs. 1. here's information about available printers, 2. here's the error

Unless others want to weigh in, I'll leave it up to you; either way is fine.